mergify_engine/merge_train.py

```diff
@@ -0,0 +1,510 @@
+            summary = "The pull request is going to be tested soon"
+        else:
+            summary = f"You can follow the pull request testing here: {tmp_pull_ctxt.pull['html_url']}"
+        summary += queue_summary
+
+        original_ctxt.log.info(
+            "pull request train car status update",
+            status=status,
+            title=title,
+            summary=summary,
+        )
+        check_api.set_check_run(
+            original_ctxt,
+            constants.MERGE_QUEUE_SUMMARY_NAME,
+            check_api.Result(
+                status,
+                title=title,
+                summary=summary,
+            ),
+        )
+
+        asyncio.run(github_events.send_refresh(original_ctxt.pull))
+
+
+@dataclasses.dataclass
+class Train:
+    client: github.GithubInstallationClient
+    redis: redis.Redis
+    owner_id: int
+    owner_login: str
+    repo_id: int
+    repo_name: str
+    ref: str
+    max_size: int = 5  # Allow to be configured
+
+    # Stored in redis
+    _cars: typing.List[TrainCar] = dataclasses.field(default_factory=list)
+    _waiting_pulls: typing.List[WaitingPull] = dataclasses.field(default_factory=list)
+    _current_base_sha: typing.Optional[github_types.SHAType] = dataclasses.field(
+        default=None
+    )
+
+    class Serialized(typing.TypedDict):
+        cars: typing.List[TrainCar.Serialized]
+        waiting_pulls: typing.List[WaitingPull]
+        current_base_sha: typing.Optional[github_types.SHAType]
+
+    def _get_train_queue_key(self) -> str:
+        return f"merge-train~{self.owner_id}~{self.repo_id}~{self.ref}"
+
+    def __post_init__(self):
+        train_raw = self.redis.get(self._get_train_queue_key())
+        if train_raw:
+            train = self.Serialized(json.loads(train_raw))
+            self._waiting_pulls = [WaitingPull(*wp) for wp in train["waiting_pulls"]]
+            self._current_base_sha = train["current_base_sha"]
+            self._cars = [TrainCar.deserialized(self, c) for c in train["cars"]]
+
+    @property
+    def log(self):
+        return daiquiri.getLogger(
+            __name__,
+            gh_owner=self.owner_login,
+            gh_repo=self.repo_name,
+            gh_branch=self.ref,
+            max_size=self.max_size,
+            train_cars=[c.pull_request_number for c in self._cars],
+            train_waiting_pulls=[wp.pull_request_number for wp in self._waiting_pulls],
+        )
+
+    def _save(self):
+        prepared = self.Serialized(
+            waiting_pulls=self._waiting_pulls,
+            current_base_sha=self._current_base_sha,
+            cars=[c.serialized() for c in self._cars],
+        )
+        raw = json.dumps(prepared)
+        self.redis.set(self._get_train_queue_key(), raw)
+
+    def get_car_by_tmp_pull(self, ctxt: context.Context) -> typing.Optional[TrainCar]:
+        for car in self._cars:
+            if car.tmp_pull_request_number == ctxt.pull["number"]:
+                return car
+        return None
```

Comment:
ok but now you can do:
```suggestion
return first.first(car for car in self._cars if car.tmp_pull_request_number == ctxt.pull["number"])
```
😉

Reply:
This method should go in stdlib !